tests/test_base.py

```diff
@@ -44,17 +46,17 @@ def test_pipeline(self):
 
         df = pd.DataFrame({"A": np.arange(1, 100), "B": np.arange(1, 100)})
 
-        params = {"df": df, "col_1": "A", "col_2": "B", "test_size": 0.2}
+        params = {"read":{"df": df.to_dict()}, "times_two":{"col_1": "A"}, "squared":{"col_2": "B"},"split": {"test_size": 0.2}}
```

Comment:
I think the data frame should be provided separately as a 3rd parameter to the `Pipeline` constructor. Then inside the constructor, we can set `params["df"] = df`. 

And while saving to a config file we can remove the `df` key and then save. In fact in `save_config`, we can remove any key with the value of type data frame. While reading, ensure that `df` key is not present and if present check if it's a valid file path.

Reply:
So, should we just remove the ```df``` if it is provided in the params object or config file?

Reply:
Yes. Let config file only save path to the dataset if provided. Primarily dataset should be passed as the 3rd parameter, and it will override any value passed in params or read from the config file.